File: MIDTERM-LAB-1/Javascript/dataQualityAudit.js
const fs = require('fs');
const readline = require('readline');

// ==========================
// READLINE SETUP
// ==========================
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// ==========================
// ASK FILE PATH
// ==========================
async function askFilePath() {
    while (true) {
        const path = await new Promise(resolve =>
            rl.question("Enter dataset file path: ", resolve)
        );

        const trimmed = path.trim();

        if (!fs.existsSync(trimmed)) {
            console.log("Error: File does not exist.\n");
            continue;
        }

        if (!fs.lstatSync(trimmed).isFile()) {
            console.log("Error: Path is not a file.\n");
            continue;
        }

        try {
            fs.accessSync(trimmed, fs.constants.R_OK);
        } catch {
            console.log("Error: File is not readable.\n");
            continue;
        }

        if (!trimmed.toLowerCase().endsWith(".csv")) {
            console.log("Error: File is not a CSV.\n");
            continue;
        }

        console.log("File validated successfully.\n");
        return trimmed;
    }
}

// ==========================
// LOAD DATASET
// ==========================
function loadDataset(filePath) {
    try {
        const rawData = fs.readFileSync(filePath, 'utf8');
        const lines = rawData
            .split(/\r?\n/)
            .filter(line => line.trim() !== "");

        if (lines.length < 2)
            throw new Error("Dataset has no data or only header.");

        return lines;
    } catch (err) {
        throw new Error("Error reading file: " + err.message);
    }
}

// ==========================
// CSV
// ==========================
function splitCsvLine(line) {
    if (!line) return [];

    const columns = [];
    let value = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const char = line[i];

        if (char === '"') {
            inQuotes = !inQuotes;
        } else if (char === ',' && !inQuotes) {
            columns.push(value);
            value = "";
        } else {
            value += char;
        }
    }

    columns.push(value);

    return columns.map(v => v.trim());
}

// ==========================
// STRICT DATE VALIDATION
// ==========================
function isValidDateStrict(dateStr) {
    if (!/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
        return false;
    }

    const [year, month, day] = dateStr.split("-").map(Number);
    const date = new Date(year, month - 1, day);

    return (
        date.getFullYear() === year &&
        date.getMonth() === month - 1 &&
        date.getDate() === day
    );
}

// ==========================
// ANALYZE DATA
// ==========================
function analyzeData(lines) {
    const headers = splitCsvLine(lines[0]);
    const records = lines.slice(1);

    let missingValues = 0;
    let negativeSales = 0;
    let invalidDates = 0;
    let duplicateRecords = 0;

    const seenRecords = new Set();

    const salesCol = headers.findIndex(h => /sales/i.test(h));
    const dateCol = headers.findIndex(h => /date/i.test(h));

    for (const line of records) {
        const columns = splitCsvLine(line);

        // ==========================
        // COUNT MISSING VALUES
        // ==========================
        for (let i = 0; i < headers.length; i++) {
            const value = columns[i];

            if (value === undefined || value.trim() === "") {
                missingValues++;
            }
        }

        // ==========================
        // DUPLICATE CHECK
        // ==========================
        const recordKey = columns.join("|");

        if (seenRecords.has(recordKey)) {
            duplicateRecords++;
        } else {
            seenRecords.add(recordKey);
        }

        // ==========================
        // NEGATIVE SALES CHECK
        // ==========================
        if (salesCol >= 0) {
            const val = columns[salesCol];
            const num = parseFloat(val);

            if (!isNaN(num) && num < 0) {
                negativeSales++;
            }
        }

        // ==========================
        // INVALID DATE CHECK
        // ==========================
        if (dateCol >= 0) {
            const val = columns[dateCol];

            if (val && val.trim() !== "") {
                if (!isValidDateStrict(val.trim())) {
                    invalidDates++;
                }
            }
        }
    }

    return {
        totalRecords: records.length,
        missingValues,
        negativeSales,
        invalidDates,
        duplicateRecords
    };
}

// ==========================
// DISPLAY REPORT
// ==========================
function displayReport(stats) {
    console.log("=======================================");
    console.log("        DATA QUALITY REPORT");
    console.log("=======================================");
    console.log("Total Records Loaded  : " + stats.totalRecords);
    console.log("Missing Values Found  : " + stats.missingValues);
    console.log("Negative Sales Found  : " + stats.negativeSales);
    console.log("Invalid Dates Found   : " + stats.invalidDates);
    console.log("Duplicate Records     : " + stats.duplicateRecords);
    console.log("=======================================");
    console.log("Audit Completed Successfully.");
}

// ==========================
// MAIN PROGRAM
// ==========================
async function main() {
    const filePath = await askFilePath();

    let lines;

    try {
        lines = loadDataset(filePath);
    } catch (err) {
        console.log(err.message);
        rl.close();
        return;
    }

    const stats = analyzeData(lines);
    displayReport(stats);

    rl.close();
}

// Run program
main();

File: MIDTERM-LAB-1/Java/DataQualityAudit.java
import java.io.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.*;

public class DataQualityAudit {

    public static void main(String[] args) {

        Scanner input = new Scanner(System.in);
        File file;

        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine();
            file = new File(path);

            if (!file.exists()) {
                System.out.println("Error: File does not exist.");
            } else if (!file.isFile()) {
                System.out.println("Error: Path is not a file.");
            } else if (!file.canRead()) {
                System.out.println("Error: File is not readable.");
            } else if (!path.toLowerCase().endsWith(".csv")) {
                System.out.println("Error: File is not in CSV format.");
            } else {
                break;
            }
            System.out.println("Please try again.\n");
        }

        System.out.println("\nLoading dataset...\n");

        int missingValues = 0;
        int negativeSales = 0;
        int invalidDates = 0;
        int duplicateRecords = 0;

        Set<String> uniqueRecords = new HashSet<>();

        int totalRecords = 0;
        int dateColumnIndex = -1;

        try (BufferedReader br = new BufferedReader(new FileReader(file))) {

            String line;

            // =========================
            // Read header properly
            // =========================
            String headerLine = br.readLine();
            if (headerLine == null) {
                System.out.println("File is empty.");
                return;
            }

            String[] headers = headerLine.split(
                ",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1
            );

            // Detect date column
            for (int i = 0; i < headers.length; i++) {
                if (headers[i].toLowerCase().contains("date")) {
                    dateColumnIndex = i;
                    break;
                }
            }

            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");

            while ((line = br.readLine()) != null) {

                if (line.trim().isEmpty()) continue;

                totalRecords++;

                String[] columns = line.split(
                    ",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)", -1
                );

                // =========================
                // Missing values
                // =========================
                for (String col : columns) {
                    if (col.trim().isEmpty()) {
                        missingValues++;
                    }
                }

                // =========================
                // Duplicate check
                // =========================
                if (!uniqueRecords.add(line)) {
                    duplicateRecords++;
                }

                // =========================
                // Negative sales
                // =========================
                for (String value : columns) {
                    try {
                        double num = Double.parseDouble(value.replace("\"", "").trim());
                        if (num < 0) {
                            negativeSales++;
                            break;
                        }
                    } catch (NumberFormatException ignored) {}
                }

                // =========================
                // Invalid Date
                // =========================
                if (dateColumnIndex >= 0 && dateColumnIndex < columns.length) {

                    String dateValue = columns[dateColumnIndex]
                            .replace("\"", "")
                            .trim();

                    if (!dateValue.isEmpty()) {
                        try {
                            LocalDate.parse(dateValue, formatter);
                        } catch (DateTimeParseException e) {
                            invalidDates++;
                        }
                    }
                }
            }

        } catch (IOException e) {
            System.out.println("Error reading file: " + e.getMessage());
            return;
        }

        // =========================
        // REPORT
        // =========================
        System.out.println("=======================================");
        System.out.println("        DATA QUALITY REPORT");
        System.out.println("=======================================");
        System.out.println("Total Records Loaded  : " + totalRecords);
        System.out.println("Missing Values Found  : " + missingValues);
        System.out.println("Negative Sales Found  : " + negativeSales);
        System.out.println("Invalid Dates Found   : " + invalidDates);
        System.out.println("Duplicate Records     : " + duplicateRecords);
        System.out.println("=======================================");
        System.out.println("Audit Completed Successfully.");
    }
}
